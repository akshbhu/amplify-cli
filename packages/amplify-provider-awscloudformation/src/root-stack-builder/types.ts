import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';

export interface AmplifyRootStackTemplate {
  Parameters?: Record<string, cdk.CfnParameter>;
  Resources?: AmplifyRootStackResource;
  Outputs?: Record<string, cdk.CfnOutput>;
  Mappings?: Record<string, cdk.CfnMapping>;
  Conditions?: Record<string, cdk.CfnCondition>;
}

export interface AmplifyRootStackResource {
  deploymentBucket?: s3.CfnBucket;
  authRole?: iam.CfnRole;
  unauthRole?: iam.CfnRole;
}
