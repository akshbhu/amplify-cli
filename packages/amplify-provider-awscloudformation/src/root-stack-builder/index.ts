export { AmplifyRootStackTransform, CommandType, ResourceConfig, RootStackTransformOptions } from './root-stack-transform';
export { AmplifyRootStackTemplate, AmplifyRootStackResource } from './types';
