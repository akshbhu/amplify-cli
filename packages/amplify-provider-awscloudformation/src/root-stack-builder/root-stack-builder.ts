import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';
import * as path from 'path';
import { prepareApp } from '@aws-cdk/core/lib/private/prepare-app';
import { AmplifyRootStackResourceProps, AmplifyRootStackTemplateProps } from './types';
import { JSONUtilities, pathManager } from 'amplify-cli-core';
import * as constants from '../constants';
import { Template } from 'cloudform-types';

const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'Root Stack for AWS Amplify CLI';

export class AmplifyRootStack extends cdk.Stack {
  public templateObj: AmplifyRootStackTemplateProps;
  constructor(scope: cdk.Construct, id: string) {
    super(scope, id);
    this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
    this.templateOptions.description = ROOT_CFN_DESCRIPTION;
    const CfnParameters: { [k: string]: cdk.CfnParameter } = {};

    CfnParameters.deploymentBucketName = new cdk.CfnParameter(this, 'deploymentBucketName', {
      type: 'String',
      description: 'Name of the common deployment bucket provided by the parent stack',
      default: 'DeploymentBucket',
    });

    CfnParameters.authRoleName = new cdk.CfnParameter(this, 'authRoleName', {
      type: 'String',
      default: 'AuthRoleName',
    });

    CfnParameters.unauthRoleName = new cdk.CfnParameter(this, 'unauthRoleName', {
      type: 'String',
      default: 'UnauthRoleName',
    });
    // Reources

    let CfnResources: AmplifyRootStackResourceProps = {};
    CfnResources.DeploymentBucket = new s3.CfnBucket(this, 'DeploymentBucket', {
      bucketName: CfnParameters['deploymentBucketName'].valueAsString,
    });

    CfnResources.DeploymentBucket.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN);

    CfnResources.authRole = new iam.CfnRole(this, 'AuthRole', {
      roleName: CfnParameters['authRoleName'].valueAsString,
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Sid: '',
            Effect: 'Deny',
            Principal: {
              Federated: 'cognito-identity.amazonaws.com',
            },
            Action: 'sts:AssumeRoleWithWebIdentity',
          },
        ],
      },
    });

    CfnResources.unauthRole = new iam.CfnRole(this, 'UnauthRole', {
      roleName: CfnParameters['unauthRoleName'].valueAsString,
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Sid: '',
            Effect: 'Deny',
            Principal: {
              Federated: 'cognito-identity.amazonaws.com',
            },
            Action: 'sts:AssumeRoleWithWebIdentity',
          },
        ],
      },
    });

    const cfnOutputs: { [k: string]: cdk.CfnOutput } = {};
    cfnOutputs.Region = new cdk.CfnOutput(this, 'Region', {
      description: 'CloudFormation provider root stack Region',
      value: cdk.Fn.ref('AWS::Region'),
      exportName: cdk.Fn.sub('${AWS::StackName}-Region'),
    });

    cfnOutputs.stackName = new cdk.CfnOutput(this, 'StackName', {
      description: 'CloudFormation provider root stack ID',
      value: cdk.Fn.ref('AWS::StackName'),
      exportName: cdk.Fn.sub('${AWS::StackName}-StackName'),
    });

    cfnOutputs.stackId = new cdk.CfnOutput(this, 'StackId', {
      description: 'CloudFormation provider root stack name',
      value: cdk.Fn.ref('AWS::StackId'),
      exportName: cdk.Fn.sub('${AWS::StackName}-StackId'),
    });

    cfnOutputs.deploymentBucketName = new cdk.CfnOutput(this, 'DeploymentBucketName', {
      description: 'CloudFormation provider root stack deployment bucket name',
      value: CfnParameters.deploymentBucketName.valueAsString,
      exportName: cdk.Fn.sub('${AWS::StackName}-DeploymentBucketName'),
    });

    cfnOutputs.authRoleArn = new cdk.CfnOutput(this, 'AuthRoleArn', {
      value: cdk.Fn.getAtt('AuthRole', 'Arn').toString(),
    });

    cfnOutputs.unauthRoleArn = new cdk.CfnOutput(this, 'UnauthRoleArn', {
      value: cdk.Fn.getAtt('UnauthRole', 'Arn').toString(),
    });

    cfnOutputs.authRoleName = new cdk.CfnOutput(this, 'AuthRoleName', {
      value: CfnParameters.authRoleName.valueAsString,
    });

    cfnOutputs.unauthRoleName = new cdk.CfnOutput(this, 'UnauthRoleName', {
      value: CfnParameters.unauthRoleName.valueAsString,
    });

    this.templateObj = { ...CfnParameters, ...cfnOutputs };
    this.templateObj.Resources = { ...CfnResources };
  }

  toCloudFormation() {
    prepareApp(this);
    return this._toCloudFormation();
  }
}

export enum CommandType {
  'PUSH',
  'INIT',
}

type RootStackOptions = {
  stackName: string;
  cfnmodifier?: Function;
  rootStackFileName: string;
  event: CommandType;
};

export const generateRootStackTemplate = async (props: RootStackOptions): Promise<Template> => {
  const stack = new AmplifyRootStack(undefined as any, 'Amplify');
  if (props.event === CommandType.INIT) {
    // no override required
    //apply init modifiers if any
  }

  if (props.event === CommandType.PUSH) {
    // apply override here during push
  }

  const cfnRootStack: Template = stack.toCloudFormation();
  await props.cfnmodifier(cfnRootStack);
  JSONUtilities.writeJson('/Users/akz/workspace/projects/init_check6/template.yml', cfnRootStack);
  return cfnRootStack;
};
