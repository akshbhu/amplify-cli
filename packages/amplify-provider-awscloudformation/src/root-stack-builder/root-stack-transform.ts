import * as path from 'path';
import { AmplifyRootStackTemplate } from './types';
import { JSONUtilities, pathManager } from 'amplify-cli-core';
import { Template } from 'cloudform-types';
import { AmplifyRootStack, AmplifyRootStackOutputs } from './root-stack-builder';
import { RootStackSythesizer } from './stackSynthesizer';
import { App } from '@aws-cdk/core';

export enum CommandType {
  'PUSH',
  'INIT',
}

type RootStackOptions = {
  rootStackFileName: string;
  event: CommandType;
  overrideDir?: string;
};

export type RootStackConfig = {
  stackFileName: string;
};

export type ResourceConfig = {
  resourceName: string;
  categoryName: string;
  stackFileName: string;
};

export interface RootStackTransformOptions {
  //inputParser: IInputParser[];  // not needed for root stack
  resourceConfig: RootStackConfig;
}

export interface DeploymentOptions {
  templateStack: Template;
}

export class AmplifyRootStackTransform {
  private app: App | undefined;
  private _rootTemplateObj: AmplifyRootStackTemplate; // Props to modify Root stack data
  private _resourceConfig: RootStackConfig; // Config about resource to override (in this case only root stack)
  private _rootStackOptions: RootStackOptions; // options to help generate  cfn template
  private _command: CommandType;
  private _synthesizer: RootStackSythesizer;
  private _synthesizerOutputs: RootStackSythesizer;

  constructor(options: RootStackTransformOptions, command: CommandType) {
    this._resourceConfig = options.resourceConfig;
    this._command = command;
    this._rootTemplateObj = {};
    this._synthesizer = new RootStackSythesizer();
    this._synthesizerOutputs = new RootStackSythesizer();
  }

  public async transform(): Promise<Template> {
    // parse Input data
    this.app = new App();
    this._rootStackOptions = await this.getInput(); // get RootStackOptions from cli-inputs.json (not required for root Stack)

    // generate cfn Constructs and AmplifyRootStackTemplate object to get overridden
    await this.generateRootStackTemplate();

    // apply override on Amplify Object having CDK Constructs for Root Stack
    await this.applyOverride();

    // generate CFN template
    const template: Template = await this.synthesizeTemplates();

    // save stack
    if (this._command === CommandType.PUSH) {
      await this.deployOverrideStacksToDisk({
        templateStack: template,
      });
    }
    return template;
  }

  private applyOverride = async () => {
    if (this._rootStackOptions.event === CommandType.PUSH) {
      // add check here to see if the override folder is present or not
      const projectPath = pathManager.findProjectRoot();
      const overridePath = pathManager.getRootOverrideDirPath(projectPath);
      const fn = await import(path.join(overridePath, 'build', 'override.js'));
      if (typeof fn.overrideProps === 'function' && fn.overrideProps != null) {
        this._rootTemplateObj = fn.overrideProps(this._rootTemplateObj);
      } else {
        console.log('There is no override setup yet for Root Stack. To enable override : Run amplify override root');
      }
    }
  };
  /**
   *
   * @returns Object required to generate Stack using cdk
   */
  private getInput = async (): Promise<RootStackOptions> => {
    if (this._command === CommandType.INIT) {
      const buildConfig: RootStackOptions = {
        event: this._command,
        rootStackFileName: this._resourceConfig.stackFileName,
      };
      return buildConfig;
    } else {
      const projectPath = pathManager.findProjectRoot();
      const buildConfig: RootStackOptions = {
        event: this._command,
        rootStackFileName: this._resourceConfig.stackFileName,
        overrideDir: pathManager.getRootOverrideDirPath(projectPath),
      };
      return buildConfig;
    }
  };

  /**
   * Generates Root stack Template
   * @returns CFN Template
   */
  private generateRootStackTemplate = async () => {
    new AmplifyRootStack(this.app, 'AmplifyRootStack', { synthesizer: this._synthesizer }, this._rootTemplateObj);
    new AmplifyRootStackOutputs(this.app, 'AmplifyRootStackOutputs', { synthesizer: this._synthesizerOutputs }, this._rootTemplateObj);
  };

  /**
   *
   * @returns return CFN templates sunthesized by app
   */
  private synthesizeTemplates = async (): Promise<Template> => {
    this.app?.synth({ force: true, skipValidation: true });
    console.log(this._rootTemplateObj);
    const templates = this._synthesizer.collectStacks();
    const templatesOutput = this._synthesizerOutputs.collectStacks();
    const cfnRootStack: Template = templates.get('AmplifyRootStack');
    const cfnRootStackOutputs: Template = templatesOutput.get('AmplifyRootStackOutputs');
    Object.assign(cfnRootStack.Outputs, cfnRootStackOutputs.Outputs);
    return cfnRootStack;
  };

  private deployOverrideStacksToDisk = async (props: DeploymentOptions) => {
    const rootFilePath = path.join(pathManager.getBackendDirPath(), 'awscloudformation', 'build', this._resourceConfig.stackFileName);
    JSONUtilities.writeJson(rootFilePath, props.templateStack);
  };
}
