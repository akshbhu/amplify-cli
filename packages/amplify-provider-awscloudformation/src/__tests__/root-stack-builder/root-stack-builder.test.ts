import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';

import * as rootStack from '../../root-stack-builder/root-stack-builder';

test('Generated rootstack template', () => {
  const stack = new rootStack.AmplifyRootStack(null, 'rootStack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
