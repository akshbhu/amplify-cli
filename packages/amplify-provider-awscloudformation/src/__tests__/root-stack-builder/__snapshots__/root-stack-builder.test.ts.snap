// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated rootstack template 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "Root Stack for AWS Amplify CLI",
  "Outputs": Object {
    "AuthRoleArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "AuthRole",
          "Arn",
        ],
      },
    },
    "AuthRoleName": Object {
      "Value": Object {
        "Ref": "authRoleName",
      },
    },
    "DeploymentBucketName": Object {
      "Description": "CloudFormation provider root stack deployment bucket name",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-DeploymentBucketName",
        },
      },
      "Value": Object {
        "Ref": "deploymentBucketName",
      },
    },
    "Region": Object {
      "Description": "CloudFormation provider root stack Region",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-Region",
        },
      },
      "Value": Object {
        "Ref": "AWS::Region",
      },
    },
    "StackId": Object {
      "Description": "CloudFormation provider root stack name",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-StackId",
        },
      },
      "Value": Object {
        "Ref": "AWS::StackId",
      },
    },
    "StackName": Object {
      "Description": "CloudFormation provider root stack ID",
      "Export": Object {
        "Name": Object {
          "Fn::Sub": "\${AWS::StackName}-StackName",
        },
      },
      "Value": Object {
        "Ref": "AWS::StackName",
      },
    },
    "UnauthRoleArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "UnauthRole",
          "Arn",
        ],
      },
    },
    "UnauthRoleName": Object {
      "Value": Object {
        "Ref": "unauthRoleName",
      },
    },
  },
  "Parameters": Object {
    "authRoleName": Object {
      "Default": "AuthRoleName",
      "Type": "String",
    },
    "deploymentBucketName": Object {
      "Default": "DeploymentBucket",
      "Description": "Name of the common deployment bucket provided by the parent stack",
      "Type": "String",
    },
    "unauthRoleName": Object {
      "Default": "UnauthRoleName",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AuthRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Effect": "Deny",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
              "Sid": "",
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": Object {
          "Ref": "authRoleName",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "DeploymentBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": Object {
          "Ref": "deploymentBucketName",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "UnauthRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Effect": "Deny",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
              "Sid": "",
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": Object {
          "Ref": "unauthRoleName",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
